File: newton_method.py
import numpy as np
from scipy.special import digamma
import matplotlib.pyplot as plt

## Setup Newton-Raphson Parameters
n = 10000 # sample size

alpha = 1.915
gamma = 135.158

X = np.random.gamma(alpha, gamma, n) # generate a gamma sample of size n
Xbar = np.mean(X) # compute the mean of the sample

logX = np.log(X) # take the log of the sample
logXbar = np.mean(logX) # take the mean of log(sample)

a = np.log(Xbar) - logXbar # constant in function phi

# phi itself
def phi(z):
    return digamma(z) - np.log(z) + a

def Ddigamma(z):
    sum = 0
    for i in range(int(1e5)):
        sum += 1 / (i + z)**2
    return sum

def Dphi(z):
    return Ddigamma(z) - 1/z

## Start N-R method
x_old = 1 # initial value

for i in range(20):
    x_new = x_old - phi(x_old) / Dphi(x_old)
    x_old = x_new

print(x_new, Xbar / x_new)

File: newton_method.R
n <- 1e6

alpha <- 2
beta <- 5

X <- rgamma(n, scale = beta, shape = alpha) # generate a gamma sample of size n
X.bar <- mean(X) # compute the mean of the sample
X.log <- log(X) # take the log of the sample
X.log.bar <- mean(X.log) # take the mean of log(sample)

a <- log(X.bar) - X.log.bar # constant in function phi

# phi itself
phi <- function(z, a) return(digamma(z) - log(z) + a)

phi.diff <- function(z) return(trigamma(z) - 1/z)

## Start N-R method
iter <- 15
x.old <- 1  # initial value
comp.alpha <- numeric(length = iter); comp.alpha[1] <- x.old # vector of estimated alpha

## Newton-Raphson method

for(i in 2:iter) {
  x.new <- x.old - phi(x.old, a) / phi.diff(x.old)
  x.old <- x.new
  comp.alpha[i] <- x.new
}

comp.beta <- X.bar / comp.alpha # vector of estimated gamma
error <- (comp.alpha - alpha) ^ 2 + (comp.beta - beta) ^ 2 # computing the error

comp.alpha[iter]
comp.beta[iter]
error[iter]
min(error)
